File: Python/DSA/n_queens.py
""" Solution to the N-Queen problem using Backtracking """

config_count = 0


def print_board(board):
    global config_count
    config_count += 1
    print(f"CONFIGURATION #{config_count}\n")

    queen_count = 1
    for row in board:
        for i in range(len(row)):
            if row[i] == 1:
                print(f"Q{queen_count}\t", end="")
                queen_count += 1
            else:
                print(f"{row[i]}\t", end="")
        print()
    print()


def is_safe(row, col, board, columns):
    x, y = row, col

    # Checking along rows
    while y >= 0:
        if board[x][y] == 1:
            return False
        y -= 1

    # Refresh the values
    x, y = row, col

    # Checking for the upper-left diagonal
    while x >= 0 and y >= 0:
        if board[x][y] == 1:
            return False
        x -= 1
        y -= 1

    # Refresh the values
    x, y = row, col

    # Checking for the lower-left diagonal
    while x < columns and y >= 0:
        if board[x][y] == 1:
            return False
        x += 1
        y -= 1

    return True


def n_queens_util(col, board, columns):
    if col == n:
        print_board(board)
        return

    for row in range(n):
        if is_safe(row, col, board, columns):
            board[row][col] = 1
            n_queens_util(col + 1, board, n)
            board[row][col] = 0


def n_queens(queens):
    board = [[0] * n for _ in range(queens)]
    n_queens_util(0, board, queens)


if __name__ == "__main__":
    n = int(input("Enter the number of Queens: "))
    print(f"\nThe possible configurations for the {n}-Queen problem are:\n")

    n_queens(n)

    print(f"Total number of configurations = {config_count}")

File: Java/DSA/GfG/Backtracking/NQueens.java
package GfG.Backtracking;
/* 
GfG problem - "N-Queen Problem" solution
Link : https://www.geeksforgeeks.org/problems/n-queen-problem0315/1
 */

import java.util.Scanner;
import java.util.ArrayList;

public class NQueens {
    private static void solve(int col, int n, ArrayList<Integer> board, boolean[] rows,
            boolean[] du, boolean[] dl, ArrayList<ArrayList<Integer>> res) {
        if (col > n) {
            // Saving the solution configuration
            res.add(new ArrayList<>(board));
            return;
        }

        for (int row = 1; row <= n; row++) {
            if (!rows[row] && !du[row - col + n] && !dl[row + col]) {
                // Step 1 : Place the Queen
                rows[row] = du[row - col + n] = dl[row + col] = true;
                board.add(row);

                // Step 2 : Onto the NEXT column
                solve(col + 1, n, board, rows, du, dl, res);

                // Step 3: Backtrack (remove the Queen)
                board.remove(board.size() - 1);
                rows[row] = du[row - col + n] = dl[row + col] = false;
            }
        }
    }

    public static ArrayList<ArrayList<Integer>> nQueens(int n) {
        ArrayList<ArrayList<Integer>> res = new ArrayList<>();
        ArrayList<Integer> board = new ArrayList<>();

        // Representing the rows : 1-based index
        boolean[] rows = new boolean[n + 1];

        // Representing the UPPER diagonal positions
        boolean[] du = new boolean[2 * n + 1];

        // Representing the LOWER diagonal positions
        boolean[] dl = new boolean[2 * n + 1];

        solve(1, n, board, rows, du, dl, res);
        return res;
    }

    public static void main(String[] args) {
        int n;
        ArrayList<ArrayList<Integer>> ans = new ArrayList<>();

        try (Scanner sc = new Scanner(System.in)) {
            System.out.print("\nEnter the number of queens : ");
            n = sc.nextInt();
        }
        ans = nQueens(n);
        System.out.print("\nPossible configurations are : < ");
        for (ArrayList<Integer> config : ans)
            System.out.print(config + " ");
        System.out.println(">\n");
    }
}
